renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -514,6 +539,550 @@ struct D3D12PixelHistoryCallback : public D3D12ActionCallback
+    if(it != m_PipeCache.end())
+      return it->second;
+
+    WrappedID3D12PipelineState *origPSO =
+        m_pDevice->GetResourceManager()->GetCurrentAs<WrappedID3D12PipelineState>(state.pipe);
+    if(origPSO == NULL)
+    {
+      RDCERR("Failed to retrieve original PSO for pixel history.");
+      return NULL;
+    }
+
+    D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC pipeDesc;
+    origPSO->Fill(pipeDesc);
+
+    ModifyPSOForStencilIncrement(eid, pipeDesc, true);
+
+    bool dxil =
+        DXBC::DXBCContainer::CheckForDXIL(pipeDesc.VS.pShaderBytecode, pipeDesc.VS.BytecodeLength);
+
+    ID3DBlob *psBlob =
+        m_pDevice->GetShaderCache()->MakeFixedColShader(D3D12ShaderCache::HIGHLIGHT, dxil);
+    if(psBlob == NULL)
+    {
+      RDCERR("Failed to create fixed color shader for pixel history.");
+      return NULL;
+    }
+
+    pipeDesc.PS.pShaderBytecode = psBlob->GetBufferPointer();
+    pipeDesc.PS.BytecodeLength = psBlob->GetBufferSize();
+
+    ID3D12PipelineState *pso = NULL;
+    HRESULT hr = m_pDevice->CreatePipeState(pipeDesc, &pso);
+    if(FAILED(hr))
+    {
+      RDCERR("Failed to create PSO for pixel history.");
+      SAFE_RELEASE(psBlob);
+      return NULL;
+    }
+
+    m_PipeCache.insert(std::make_pair(state.pipe, pso));
+    SAFE_RELEASE(psBlob);
+    return pso;
+  }
+
+private:
+  D3D12RenderState m_SavedState;
+  std::map<ResourceId, ID3D12PipelineState *> m_PipeCache;
+  rdcarray<uint32_t> m_Events;
+
+  // Key is event ID, and value is an index of where the occlusion result.
+  std::map<uint32_t, uint32_t> m_OcclusionQueries;
+  rdcarray<uint64_t> m_OcclusionResults;
+};
+
+struct D3D12ColorAndStencilCallback : public D3D12PixelHistoryCallback
+{
+  D3D12ColorAndStencilCallback(WrappedID3D12Device *device, D3D12PixelHistoryShaderCache *shaderCache,
+                               const D3D12PixelHistoryCallbackInfo &callbackInfo,
+                               const rdcarray<uint32_t> &events)
+      : D3D12PixelHistoryCallback(device, shaderCache, callbackInfo, NULL), m_Events(events)
+  {
+  }
+
+  ~D3D12ColorAndStencilCallback()
+  {
+    for(auto it = m_PipeCache.begin(); it != m_PipeCache.end(); ++it)
+    {
+      SAFE_RELEASE(it->second.fixedShaderStencil);
+      SAFE_RELEASE(it->second.originalShaderStencil);
+    }
+  }
+
+  void PreDraw(uint32_t eid, ID3D12GraphicsCommandListX *cmd)
+  {
+    if(!m_Events.contains(eid))
+      return;
+
+    D3D12MarkerRegion::Set(cmd, StringFormat::Fmt("Replaying event %u", eid));
+
+    m_SavedState = m_pDevice->GetQueue()->GetCommandData()->GetCurRenderState();
+    D3D12RenderState &pipeState = m_pDevice->GetQueue()->GetCommandData()->GetCurRenderState();
+
+    // Get pre-modification values
+    size_t storeOffset = m_EventIndices.size() * sizeof(D3D12EventInfo);
+
+    CopyPixel(eid, cmd, storeOffset + offsetof(struct D3D12EventInfo, premod));
+
+    {
+      D3D12PipelineReplacements replacements = GetPipelineReplacements(eid, pipeState);
```

Comment:
I think worth a TODO perhaps a general one at the file scope about the current support for pixel history for multiple render targets.

There is some support ie. `GetPixelHistoryRenderTargetIndex`, there also appears to be places which are assuming target index of 0 in particular for the replacement shaders.

Reply:
That's reasonable, adding a TODO at the top of the file: "Broadly speaking, support for pixel history with multiple render targets bound needs more test coverage and testing to ensure proper implementation."